src/validators/authorizer.rs

```diff
@@ -0,0 +1,405 @@
+use crate::{
+	apis::jwks_api::{Jwks, JwksModel},
+	clerk::Clerk,
+};
+use jsonwebtoken::{decode, decode_header, errors::Error as jwtError, Algorithm, DecodingKey, Header, Validation};
+use std::{error::Error, fmt};
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct ClerkJwt {
+	pub azp: String,
+	pub exp: i32,
+	pub iat: i32,
+	pub iss: String,
+	pub nbf: i32,
+	pub sid: String,
+	pub sub: String,
+}
+
+pub trait ClerkRequest {
+	fn get_header(&self, key: &str) -> Option<String>;
+	fn get_cookie(&self, key: &str) -> Option<String>;
+}
+
+#[derive(Debug)]
+pub enum ClerkError {
+	Unauthorized(String),
+	InternalServerError(String),
+}
+
+impl fmt::Display for ClerkError {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		match self {
+			ClerkError::Unauthorized(msg) => write!(f, "Unauthorized: {}", msg),
+			ClerkError::InternalServerError(msg) => write!(f, "Internal Server Error: {}", msg),
+		}
+	}
+}
+
+impl Error for ClerkError {}
+
+#[derive(Clone)]
+pub struct ClerkAuthorizer {
+	clerk_client: Clerk,
+	validate_session_cookie: bool,
+}
+
+impl ClerkAuthorizer {
+	/// Creates a Clerk authorizer
+	pub fn new(clerk_client: Clerk, validate_session_cookie: bool) -> Self {
+		Self {
+			clerk_client,
+			validate_session_cookie,
+		}
+	}
+
+	/// Authorizes a service request against the Clerk auth provider
+	pub async fn authorize<T>(&self, request: &T) -> Result<ClerkJwt, ClerkError>
+	where
+		T: ClerkRequest,
+	{
+		let jwks = match Jwks::get_jwks(&self.clerk_client).await {
+			Ok(val) => val,
+			Err(_) => return Err(ClerkError::InternalServerError(String::from("Error: Could not fetch JWKS!"))),
+		};
+
+		let access_token: String = match request.get_header("Authorization") {
+			Some(val) => val.to_string().replace("Bearer ", ""),
+			None => match self.validate_session_cookie {
+				true => match request.get_cookie("__session") {
+					Some(cookie) => cookie.to_string(),
+					None => {
+						return Err(ClerkError::Unauthorized(String::from(
+							"Error: No Authorization header or session cookie found on the request payload!",
+						)))
+					}
+				},
+				false => {
+					return Err(ClerkError::Unauthorized(String::from(
+						"Error: No Authorization header found on the request payload!",
+					)))
+				}
+			},
+		};
+
+		match validate_jwt(&access_token, jwks) {
+			Ok(val) => Ok(val),
+			Err(_) => return Err(ClerkError::Unauthorized(String::from("Error: Invalid JWT!"))),
+		}
+	}
+}
+
+/// Validates a jwt token using a jwks
+fn validate_jwt(token: &str, jwks: JwksModel) -> Result<ClerkJwt, ()> {
```

Comment:
This function needs to be exported - some users interface with this directly. I think its better if we revert this function back to what it was before (always returning `true` or `false`)

```rust
pub fn validate_jwt(token: &str, jwks: JwksModel) -> Result<(bool, ClerkJwt), bool> {
```

Reply:
Done.